### ISSUE TITLE
Unclear whether `.call` can be implicitly torn off when supplying type arguments

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#1651]
Traditionally we have allowed the user to elide `.call` in a lot of circumstances, so for example:
```dart
class C {
  void call() {}
}
g(void Function()) { ... }
f(C c) {
  c(); // Same as `c.call()`
  g(c); // Same as `g(c.call)`
}
```
The current spec for constructor tear-offs adds support for the syntax `<TypeArguments>` after a reference to a function.  Do we want to allow implicit reference to `.call` for this syntax as well?  In other words:
```dart
class D {
  void call<T>(T t) {}
}
h(D d1, D d2, bool b) {
  void Function(int) x1 = d1.call<int>; // Definitely allowed
  void Function(int) x2 = d1<int>; // Should this be allowed? (same as `d1.call<int>`)
  void Function(int) x3 = (b ? d1 : d2).call<int>; // Definitely allowed
  void Function(int) x4 = (b ? d1 : d2)<int>; // Should this be allowed? (same as `(b ? d1 : d2).call<int>`)
}
```
It's not clear to me whether the current spec draft would allow `x2` and `x4` or not.

**Comment by stereotype441** [comment]
Note: the parser work I've done so far for constructor tear-offs is flexible enough to allow any arbitrary expression to the left of `<TypeArguments>` (it just kind of falls naturally out of allowing `<TypeArguments>` as a selector).  So if we want to allow `x2` and `x4` in the example above, I believe it would be fairly easy to do so.

**Comment by stereotype441** [comment]
@lrhn @eernstg @leafpetersen FYI

**Comment by eernstg** [comment]
We could eliminate the support for callable objects entirely, but we just had that discussion recently (in connection with `super()`). For consistency, assuming that we will preserve the current support for callable objects, I think we should also include support for this (that is, implicitly inducing `.call` before `<typeArguments>` in `<primary> <selector>* <typeArguments>` when it is applied to a callable object).

Pragmatically, this might be helpful because it allows refactorings like this one:

```dart
class C { void call<X>(X x) {}}

void foo(C c) {
  void Function(int) f = c; // Old code, which may now be written as ..
  var f = c<int>;
}
```

Of course, the developer could very well write `var f = c.call<int>;` but that's just yet another little thing to remember.

**Comment by lrhn** [comment]
The current constructor tear-off proposal (revision 2.6) [explicitly *disallows*](https://github.com/dart-lang/language/blob/master/working/0216%20-%20constructor%20tearoffs/proposal.md#no-instantiated-tearing-off-function-call-methods) instantiating `call` when torn off from a receiver of type `Function` or a function type (and has done so since revision 2.4). That was introduced after we decided not to allow such instantiated function `call`-method tear-offs at the preceding language meeting, which we did after realizing that implementations don't actually support it now, and it's non-trivial to support using the current implementation approach.

(You can instantiate the tear off from a callable object, that's just a normal instance method).

**Comment by eernstg** [comment]
Right, and those errors do not conflict with support for implicitly adding `.call` on a callable object.

**Comment by leafpetersen** [comment]
> Tl;dr: Do allow `c.call<int>` and don't allow `c<int>`. Only allow `e<typeArgs>` when `e` denotes a function declaration, type declaration or instance method.

Agreed.

**Comment by eernstg** [comment]
Trying out the code from [here](https://github.com/dart-lang/language/issues/1651#issuecomment-849529916), I noted that the following is actually rejected as a compile-time error by the analyzer as well as the CFE:

```dart
int Function(int) id9 = c; // valid, equivalent to c.call, which is valid (below)
```

This is actually the consistent treatment from the following point of view:

> Callable objects are _not_ equivalent to function declarations, they are equivalent to
> function _values_. Function values cannot be instantiated

The following 3 errors are reported by the CFE, but the analyzer accepts these declarations with no errors (presumably that's because the analyzer hasn't yet implemented the error for instantiating the `call` method of a function object):

```
n027.dart:13:28: Error: A value of type 'T Function<T>(T)' can't be assigned to a variable of type 'int Function(int)'.
int Function(int) id3 = id.call;  // invalid, instantiates `call` method implicitly.
                           ^
n027.dart:16:25: Error: A value of type 'T Function<T>(T)' can't be assigned to a variable of type 'int Function(int)'.
int Function(int) id5 = idfunc; // invalid, instantiates function value, not declaration.
                        ^
n027.dart:19:32: Error: A value of type 'T Function<T>(T)' can't be assigned to a variable of type 'int Function(int)'.
int Function(int) id7 = idfunc.call; // invalid, instantiates `call` method of function.
                               ^
```

> Tl;dr: Do allow `c.call<int>` and don't allow `c<int>`.

That conclusion is now even more consistent! (.. because `c` is not supported after all, so we do _not_ have the semantic discrepancy where `c` allows instantiation implicitly, but the explicit version `c<int>` is disallowed, they are both disallowed).
